### PR TITLE
Refactor HTTP protocol generation

### PR DESCRIPTION
**Pull request by kstich**
This commit is a significant refactor of the previous iteration of
generating HTTP protocols. Most importantly, the process of building
out support for a protocol is more composible. It has been broken
down into 2 components.

1. The HttpBindingProtocolGenerator implementation, responsible for
handling or dispatching document serde and generating components for
that.
2. The DocumentShapeSerdeVisitor implementations, responsible for
generating serde functions specific to shapes contained within the
document body of a protocol. Implmenentations for serialization
and deserialization are provided.

HttpBindingProtocolGenerator implementations are responsible for
tracking which shapes need serde generated by the implementations
of DocumentShapeSerdeVisitors.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeGenerator.java [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+package software.amazon.smithy.typescript.codegen.integration;
+
+import software.amazon.smithy.model.shapes.CollectionShape;
+import software.amazon.smithy.model.shapes.MapShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+import software.amazon.smithy.typescript.codegen.integration.ProtocolGenerator.GenerationContext;
+
+/**
+ * Smithy protocol document body shape generators.
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeGenerator.java]: I wonder if we should use the word "synthesized" here to better delineate this from document types?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeGenerator.java [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+
+    /**
+     * Writes the code needed to serialize a structure in the document of a request.
+     *
+     * <p>Implementations of this method are expected to generate a function body that
+     * returns a value representing the StructureShape {@code shape} parameter that is
+     * serializable by {@code serializeDocument}.
+     *
+     * <p>The function signature for this body will have two parameters:
+     *   <ul>
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeGenerator.java]: Is this supposed to be indented?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeGenerator.java [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+public interface DocumentShapeSerdeGenerator extends HttpValueProvider {
+
+    /**
+     * Writes the code needed to serialize a structure in the document of a request.
+     *
+     * <p>Implementations of this method are expected to generate a function body that
+     * returns a value representing the StructureShape {@code shape} parameter that is
+     * serializable by {@code serializeDocument}.
+     *
+     * <p>The function signature for this body will have two parameters:
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeGenerator.java]: Should this further clarify "will have two parameters:" -> "will have two parameters available in scope:"

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeGenerator.java [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+     *   <ul>
+     *     <li>{@code input}: the type generated for the CollectionShape shape parameter.</li>
+     *     <li>{@code context}: the SerdeContext.</li>
+     *   </ul>
+     *
+     * <p>For example, this function would generate the following:
+     *
+     * <pre>{@code
+     *   return (input || []).map(entry =>
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeGenerator.java]: ideally, you don't need to indent the contents of these methods manually, right?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeGenerator.java [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+     *   <ul>
+     *     <li>{@code output}: a value representing the StructureShape shape parameter.</li>
+     *     <li>{@code context}: the SerdeContext.</li>
+     *   </ul>
+     *
+     * <p>For example, this function would generate the following:
+     *
+     * <pre>{@code
+     *   let contents: any = {
+     *     $namespace: "com.smithy.example",
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeGenerator.java]: We don't use the $namespace as of today

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeVisitor.java [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+import software.amazon.smithy.model.shapes.SetShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.ShapeVisitor;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+import software.amazon.smithy.typescript.codegen.TypeScriptWriter;
+import software.amazon.smithy.typescript.codegen.integration.ProtocolGenerator.GenerationContext;
+
+/**
+ * Visitor to generate serde functions for shapes in protocol document bodies.
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerdeVisitor.java]: More info on what this is and why you'd use it would be nice

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpDocumentBodyGenerator.java [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.Set;
+import software.amazon.smithy.model.knowledge.HttpBinding;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.typescript.codegen.integration.ProtocolGenerator.GenerationContext;
+
+/**
+ * Abstract implementation useful for all protocols that have HTTP document bodies.
+ */
+public abstract class HttpDocumentBodyGenerator implements HttpValueProvider {
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpDocumentBodyGenerator.java]: Why is this class necessary when we already have the other document body generators?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpDocumentBodyGenerator.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    /**
+     * Gets the visitor used to generate deserialization functions for shapes.
+     *
+     * @param context The generation context.
+     * @return Returns the deserialization visitor implementation.
+     */
+    protected abstract DocumentShapeDeserializationVisitor getDocumentShapeDeserializer(GenerationContext context);
+
+    /**
+     *
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpDocumentBodyGenerator.java]: Missing docs.

Why is this necessary if there's already generateSharedComponents?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerializationGenerator.java [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+import software.amazon.smithy.model.shapes.CollectionShape;
+import software.amazon.smithy.model.shapes.MapShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+
+/**
+ * Generates document body shape serializers used to synthesize a protocol
+ * document from generated types.
+ */
+public interface DocumentShapeSerializationGenerator extends DocumentShapeSerdeVisitor {
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerializationGenerator.java]: How would we serialize document types?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpDocumentBodyGenerator.java (hunk 7 above)
Comment (mtdowling) [reply smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpDocumentBodyGenerator.java]: I still don't understand the purpose of this class vs the others. Why is this necessary?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpValueProvider.java [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+import software.amazon.smithy.model.shapes.NumberShape;
+import software.amazon.smithy.model.shapes.Shape;
+import software.amazon.smithy.model.shapes.StringShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.TimestampShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+import software.amazon.smithy.model.traits.TimestampFormatTrait.Format;
+import software.amazon.smithy.typescript.codegen.integration.ProtocolGenerator.GenerationContext;
+
+interface HttpValueProvider {
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpValueProvider.java]: This could use both docs and tests

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpDocumentBodyGenerator.java [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+ * that make up those documents.
+ */
+public abstract class HttpDocumentBodyGenerator implements HttpValueProvider {
+
+    /**
+     * Gets the default content-type when a document is synthesized in the body.
+     *
+     * @return Returns the default content-type.
+     */
+    protected abstract String getDocumentContentType();
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpDocumentBodyGenerator.java]: Why is this no longer part of `HttpBindingProtocolGenerator`? Doesn't this now couple this class to HttpBindingProtocolGenerator if HttpBindingProtocolGenerator is meant to call this method?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserializationGenerator.java [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+import software.amazon.smithy.model.shapes.DocumentShape;
+import software.amazon.smithy.model.shapes.MapShape;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.model.shapes.UnionShape;
+import software.amazon.smithy.typescript.codegen.integration.ProtocolGenerator.GenerationContext;
+
+/**
+ * Generates document body shape deserializers used to synthesize generated
+ * types from a protocol document.
+ */
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserializationGenerator.java]: What's fair game to be added here? Why are some things in this interface but not others? What's the designed purpose of this interface?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java [hunk 13]
```diff
@@ -63,10 +43,13 @@
 /**
  * Abstract implementation useful for all protocols that use HTTP bindings.
  */
-public abstract class HttpBindingProtocolGenerator implements ProtocolGenerator {
+public abstract class HttpBindingProtocolGenerator implements ProtocolGenerator, HttpValueProvider {
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java]: There's a *lot* of protected methods on this class. I'm wondering if there's a way to trim this number down using composition instead.

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserVisitor.java [hunk 14]
```diff
@@ -0,0 +1,433 @@
...
+ * their loading from the body to the matching abstract method. The {@link DocumentMemberDeserVisitor}
+ * is provided to reduce the effort of this implementation by providing the default strategies
+ * for deserializing content from aggregate type members.
+ *
+ * Visitor methods for all other types will default to not generating deserialization
+ * functions. This may be overwritten by downstream implementations if the protocol requires
+ * more complex deserialization strategies for those types.
+ *
+ */
+public abstract class DocumentShapeDeserVisitor extends ShapeVisitor.Default<Void> {
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserVisitor.java]: Should this link and refer to HttpBindingProtocolGenerator to give more context on how this work in practice? And I assume this can also be used independent of HttpBindingProtocolGenerator too, right?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserVisitor.java [hunk 15]
```diff
@@ -0,0 +1,433 @@
...
+     *   <li>{@code output}: a value representing the MapShape shape parameter.</li>
+     *   <li>{@code context}: the SerdeContext.</li>
+     * </ul>
+     *
+     * <p>For example, this function would generate the following:
+     *
+     * <pre>{@code
+     * let mapParams: any = {};
+     * Object.keys(output).forEach(key => {
+     *   mapParams[key] = deserializeAws_restJson1_1Field(output[key], context);
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserVisitor.java]: Doesn't this need the [hasOwnProperty](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty) check to not send random members from the object?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserVisitor.java [hunk 16]
```diff
@@ -0,0 +1,433 @@
...
+
+    /**
+     * Generates a function for serializing the input shape, dispatching the body generation
+     * to the supplied function.
+     *
+     * @param shape The shape to generate a serializer for.
+     * @param functionBody An implementation that will generate a function body to
+     *                     serialize the shape.
+     */
+    protected void generateDeserFunction(
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserVisitor.java]: Can this be private for now at least?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java [hunk 17]
```diff
@@ -0,0 +1,440 @@
...
+ *
+ * Visitor methods for all other types will default to not generating serialization functions.
+ * This may be overwritten by downstream implementations if the protocol requires more
+ * complex serialization strategies for those types.
+ */
+public abstract class DocumentShapeSerVisitor extends ShapeVisitor.Default<Void> {
+    private GenerationContext context;
+    private String protocolName;
+
+    public DocumentShapeSerVisitor(GenerationContext context, String protocolName) {
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java]: Should we add protocolName to GenerationContext?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java [hunk 18]
```diff
@@ -0,0 +1,440 @@
...
+ * their body generation to the matching abstract method. The {@link DocumentMemberSerVisitor}
+ * is provided to reduce the effort of this implementation by providing the default strategies
+ * for serializing content from aggregate type members.
+ *
+ * Visitor methods for all other types will default to not generating serialization functions.
+ * This may be overwritten by downstream implementations if the protocol requires more
+ * complex serialization strategies for those types.
+ */
+public abstract class DocumentShapeSerVisitor extends ShapeVisitor.Default<Void> {
+    private GenerationContext context;
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java]: These should both be final

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java [hunk 19]
```diff
@@ -0,0 +1,440 @@
...
+        this.context = context;
+        this.protocolName = protocolName;
+    }
+
+    /**
+     * Gets the generation context.
+     *
+     * @return The generation context.
+     */
+    protected GenerationContext getContext() {
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java]: final?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java [hunk 20]
```diff
@@ -0,0 +1,440 @@
...
+        return context;
+    }
+
+    /**
+     * Gets the name of the protocol.
+     *
+     * @return The protocol name.
+     */
+    protected String getProtocolName() {
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java]: final?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java [hunk 21]
```diff
@@ -0,0 +1,440 @@
...
+    /**
+     * Writes the code needed to serialize a collection in the document of a request.
+     *
+     * <p>Implementations of this method are expected to generate a function body that
+     * returns a value representing the CollectionShape {@code shape} parameter that is
+     * serializable by {@code serializeInputDocument}.
+     *
+     * <p>The function signature for this body will have two parameters available in scope:
+     * <ul>
+     *   <li>{@code input}: the type generated for the CollectionShape shape parameter.</li>
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java]: Again from other comment -- these should be more specific about the TypeScript types of each of these variables

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java [hunk 22]
```diff
@@ -0,0 +1,440 @@
...
+
+    /**
+     * Generates a function for serializing the input shape, dispatching the body generation
+     * to the supplied function.
+     *
+     * @param shape The shape to generate a serializer for.
+     * @param functionBody An implementation that will generate a function body to
+     *                     serialize the shape.
+     */
+    protected void generateSerFunction(
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeSerVisitor.java]: Can this be private for now at least?

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java [hunk 23]
```diff
@@ -67,6 +54,9 @@
 
     private static final Logger LOGGER = Logger.getLogger(HttpBindingProtocolGenerator.class.getName());
 
+    private Set<Shape> documentSerializingShapes = new TreeSet<>();
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java]: both final

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java [hunk 24]
```diff
@@ -86,8 +76,33 @@ public ApplicationProtocol getApplicationProtocol() {
...
+    /**
+     * Generates deserialization functions for shapes in the passed set. These functions
+     * should return a value that can then be deserialized by the implementation of
+     * {@code deserializeDocument}. The {@link DocumentShapeDeserVisitor} and
+     * {@link DocumentMemberDeserVisitor} are provided to reduce the effort of this implementation.
+     *
+     * @param context The generation context.
+     * @param shapes The shapes to generate deserialization for.
+     */
+    protected abstract void generateDocumentShapeDeserializers(GenerationContext context, Set<Shape> shapes);
```
Comment (mtdowling) [inline smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java]: Should this and generateDocumentShapeSerializers just become a single method, `generateDocumentShapeSerdeFunctions`? It seems like them as separate methods doesn't actually help anything, does it? Or are the provided Sets totally different?

Comment (kstich) [reply smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/HttpBindingProtocolGenerator.java]: The sets themselves are likely to be quite distinct. The most obvious example is for errors, but an API may consume shapes that aren't given as output, and vice versa, and generating functions for those is unused overhead.

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentMemberSerVisitor.java [hunk 25]
```diff
@@ -0,0 +1,220 @@
...
+    public final String memberShape(MemberShape shape) {
+        throw new CodegenException("Member shapes cannot be bound to documents.");
+    }
+
+    public String getTimestampSerializedWithFormat(TimestampShape shape, Format format) {
+        switch (format) {
+            case DATE_TIME:
+                return dataSource + ".toISOString()";
+            case EPOCH_SECONDS:
+                return "Math.round(" + dataSource + ".getTime() / 1000)";
```
[the remark this answers is not in the document]
Comment (kstich) [reply smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentMemberSerVisitor.java]: `dataSource` in this context is a Date object, [where invoking `getTime` returns milliseconds](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/getTime).

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserVisitor.java (hunk 15 above)
Comment (kstich) [reply smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentShapeDeserVisitor.java]: Maps are generated as [`{ [key: string]: %s }`](https://github.com/awslabs/smithy-typescript/blob/master/smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/SymbolVisitor.java#L161), where %s is the generated type of the map value. As such, they won't have additional properties and the check shouldn't be necessary.

File: smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentMemberDeserVisitor.java [hunk 26]
```diff
@@ -0,0 +1,193 @@
...
+    }
+
+    @Override
+    public final String memberShape(MemberShape shape) {
+        throw new CodegenException("Member shapes cannot be bound to documents.");
+    }
+
+    @Override
+    public String timestampShape(TimestampShape shape) {
+        return "new Date(" + dataSource + ")";
```
[the remark this answers is not in the document]
Comment (kstich) [reply smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/integration/DocumentMemberDeserVisitor.java]: Will make this update for getOutputValue as well.
